File: Testing/tests/client/Tests/BasicCSharp/TestAbstractClass.cs
using Bridge;
using Bridge.QUnit;

namespace ClientTestLibrary
{
    class TestAbstractClass
    {
        [FileName("testAbstractClass.js")]
        abstract class A
        {
            public int Data { get; set; }
            public abstract string GetString();
        }

        [FileName("testAbstractClass.js")]
        class B : A
        {
            public override string GetString()
            {
                Data++;
                return "B";
            }
        }

        [FileName("testAbstractClass.js")]
        class C : B
        {
            public override string GetString()
            {
                Data--;
                return "C";
            }
        }

        public static void TestB(Assert assert)
        {
            assert.Expect(3);

            var b = new B();

            assert.Ok(b != null, "Instance of B created");
            assert.Equal(b.GetString(), "B", "b.GetString() = 'B'");
            assert.Equal(b.Data, 1, "b.Data = 1");
        }

        public static void TestC(Assert assert)
        {
            assert.Expect(3);

            var c = new C();

            assert.Ok(c != null, "Instance of C created");
            assert.Equal(c.GetString(), "C", "c.GetString() = 'C'");
            assert.Equal(c.Data, -1, "c.Data = -1");
        }

        public static void TestBC(Assert assert)
        {
            assert.Expect(6);

            A b = new B();

            assert.Ok(b != null, "Instance of B created as instance of A");
            assert.Equal(b.GetString(), "B", "b.GetString() = 'B'");
            assert.Equal(b.Data, 1, "b.Data = 1");

            A c = new C();
            assert.Ok(c != null, "Instance of C created as instance of A");
            assert.Equal(c.GetString(), "C", "c.GetString() = 'C'");
            assert.Equal(c.Data, -1, "c.Data = -1");
        }
    }
}

File: Testing/www/js/testAbstractClass.js
/* global Bridge */

"use strict";

Bridge.define('ClientTestLibrary.TestAbstractClass', {
    statics: {
        testB: function (assert) {
            assert.expect(3);

            var b = new ClientTestLibrary.TestAbstractClass.B();

            assert.ok(b !== null, "Instance of B created");
            assert.equal(b.getString(), "B", "b.GetString() = 'B'");
            assert.equal(b.getData(), 1, "b.Data = 1");
        },
        testC: function (assert) {
            assert.expect(3);

            var c = new ClientTestLibrary.TestAbstractClass.C();

            assert.ok(c !== null, "Instance of C created");
            assert.equal(c.getString(), "C", "c.GetString() = 'C'");
            assert.equal(c.getData(), -1, "c.Data = -1");
        },
        testBC: function (assert) {
            assert.expect(6);

            var b = new ClientTestLibrary.TestAbstractClass.B();

            assert.ok(b !== null, "Instance of B created as instance of A");
            assert.equal(b.getString(), "B", "b.GetString() = 'B'");
            assert.equal(b.getData(), 1, "b.Data = 1");

            var c = new ClientTestLibrary.TestAbstractClass.C();
            assert.ok(c !== null, "Instance of C created as instance of A");
            assert.equal(c.getString(), "C", "c.GetString() = 'C'");
            assert.equal(c.getData(), -1, "c.Data = -1");
        }
    }
});

Bridge.define('ClientTestLibrary.TestAbstractClass.A', {
    config: {
        properties: {
            Data: 0
        }
    }
});

Bridge.define('ClientTestLibrary.TestAbstractClass.B', {
    inherits: [ClientTestLibrary.TestAbstractClass.A],
    getString: function () {
        this.setData(this.getData()+1);
        return "B";
    }
});

Bridge.define('ClientTestLibrary.TestAbstractClass.C', {
    inherits: [ClientTestLibrary.TestAbstractClass.B],
    getString: function () {
        this.setData(this.getData()-1);
        return "C";
    }
});



Bridge.init();